### PR TITLE
New Resource: `azurerm_arc_machine`

### PR DESCRIPTION
**Pull request by neil-yechenwei**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description
This is the PR to support Hybrid Compute Machine. 
Given HC suggests to use arc_* for this PR, so I named it to azurerm_arc_machine.

API Spec: https://github.com/Azure/azure-rest-api-specs/blob/1ad29756bd141a47cac770140105a706d065ae1b/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2024-07-10/HybridCompute.json#L352

Overview doc: https://learn.microsoft.com/en-us/azure/azure-arc/system-center-virtual-machine-manager/overview

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New Resource: `azurerm_arc_machine`

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/hybridcompute/arc_machine_resource.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			state := ArcMachineResourceModel{}
+			if model := resp.Model; model != nil {
+				state.Name = id.MachineName
+				state.ResourceGroupName = id.ResourceGroupName
+				state.Location = location.Normalize(model.Location)
+				state.Kind = string(pointer.From(model.Kind))
+			}
```
Comment (ms-zhenhua) [inline internal/services/hybridcompute/arc_machine_resource.go:132]:
```suggestion
                        state.Name = id.MachineName
		        state.ResourceGroupName = id.ResourceGroupName
			if model := resp.Model; model != nil {
				state.Location = location.Normalize(model.Location)
				state.Kind = string(pointer.From(model.Kind))
			}
```

File: internal/services/hybridcompute/arc_machine_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+func (r ArcMachineResource) Exists(ctx context.Context, clients *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
+	id, err := machines.ParseMachineID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.HybridCompute.HybridComputeClient_v2024_05_20_preview.Machines.Get(ctx, *id, machines.DefaultGetOperationOptions())
+	if err != nil {
+		return nil, fmt.Errorf("reading %s: %+v", *id, err)
```
Comment (ms-zhenhua) [inline internal/services/hybridcompute/arc_machine_resource_test.go:59]:
```suggestion
		return nil, fmt.Errorf("retrieving %s: %+v", *id, err)
```

File: internal/services/hybridcompute/arc_machine_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	return utils.Bool(resp.Model != nil), nil
```
Comment (ms-zhenhua) [inline internal/services/hybridcompute/arc_machine_resource_test.go:62]:
```suggestion
	return pointer.To(resp.Model != nil), nil
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
@neil-yechenwei, Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: internal/services/hybridcompute/client/client.go [hunk 4]
```diff
@@ -10,17 +10,27 @@ import (
...
+	PrivateEndpointConnectionsClient        *privateendpointconnections.PrivateEndpointConnectionsClient
+	PrivateLinkScopesClient                 *privatelinkscopes.PrivateLinkScopesClient
 }
 
 func NewClient(o *common.ClientOptions) (*Client, error) {
+	hybridComputeClient_v2024_05_20_preview, err := hybridcompute_v2024_05_20_preview.NewClientWithBaseURI(o.Environment.ResourceManager, func(c *resourcemanager.Client) {
+		o.Configure(c, o.Authorizers.ResourceManager)
+	})
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hybridcompute/client/client.go:31]: Updated

File: internal/services/hybridcompute/arc_machine_resource.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	return machines.ValidateMachineID
+}
+
+func (r ArcMachineResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hybridcompute/arc_machine_resource.go:48]: I didn't find any validation from azure portal and service API.

File: website/docs/r/arc_machine.html.markdown [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Arc machine. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Arc Machine should exist. Changing this forces a new resource to be created.
+
+* `location` - (Required) The Azure Region where the Arc Machine should exist. Changing this forces a new resource to be created.
+
+* `kind` - (Required) The kind of the Arc Machine. Changing this forces a new resource to be created.
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply website/docs/r/arc_machine.html.markdown:39]: Updated

File: internal/services/hybridcompute/arc_machine_resource.go (hunk 1 above)
Comment (neil-yechenwei) [reply internal/services/hybridcompute/arc_machine_resource.go:132]: Updated

File: internal/services/hybridcompute/arc_machine_resource_test.go (hunk 2 above)
Comment (neil-yechenwei) [reply internal/services/hybridcompute/arc_machine_resource_test.go:59]: Updated

File: internal/services/hybridcompute/arc_machine_resource_test.go (hunk 3 above)
Comment (neil-yechenwei) [reply internal/services/hybridcompute/arc_machine_resource_test.go:62]: Updated

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , thanks for the comments. I updated PR. Please take another look.

**Review by ms-zhenhua — APPROVED** [review]

**Comment by ms-zhenhua** [comment]
@neil-yechenwei, thank you for your updates. LGTM~

**Review by katbyte — APPROVED** [review]
LGTM 🛋️

**Pull request merged**